### ISSUE TITLE
Add selector-max-class rule

### ISSUE THREAD

**Issue by jimbo** [stylelint/stylelint#2458]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Propose adding a **new option** for [`selector-max-specificity`](https://stylelint.io/user-guide/rules/selector-max-specificity/)

- Propose naming the option `granular` and accepting possible values of `true|false`
- With `granular: true`, limit individual components of specificity rather than cumulative specificity

> Which rule, if any, is this issue related to?

[`selector-max-specificity`](https://stylelint.io/user-guide/rules/selector-max-specificity/)

> What CSS is needed to reproduce this issue?

```css
.one .two .three .four .five .six .seven .eight .nine .ten {
  color: pink;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-max-specificity": "1,3,1"
  }
}
```

> Which version of stylelint are you using?

`latest` (`7.9.0`)

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

Stylelint works as expected. However, if my configuration allows for selectors to contain at least one id, I expect there to be some way to limit the number of classnames and/or tagnames.

> What actually happened (e.g. what warnings or errors you are getting)?

Without a new option, the rule is (correctly) not violated because a selector with ten classnames has a lower *cumulative* specificity than the configured maximum, `"1,3,1"`, which represents a selector with one id. As far as I know, there's currently no way to alter this behavior.

---

As it happens, I didn't discover the [rules](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) until I'd already forked the library and implemented the option. According to the rules, we're supposed to discuss the option and its name here first, but I can put up a PR whenever.

**Comment by hudochenkov** [comment]
I think it's a clear candidate for a [plugin](https://stylelint.io/developer-guide/plugins/), because this option doesn't have a clear and unambiguous finished state, and, I believe, won't be used by majority of stylelint's users.

**Comment by jeddy3** [comment]
@jimbo Thanks for taking the time to writing up a proposal.

It took me a bit to get my head around it, and I'm not sure I understand it completely yet. Am I right in thinking that the way the `selector-max-specificity` currently works is in line with [with the spec](https://drafts.csswg.org/selectors/#specificity-rules), and what you're proposing is some adaptation of that?

If so, then I agree with @hudochenkov that this is a perfect candidate for a plugin. Plugins are where language extensions live. You're `granular` name makes sense to me, and so you may want to call the plugin `stylelint-selector-max-granular-specificity`, or something similar.

If you do go down the plugin route, feel free to create a PR to add it [to the list](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/plugins.md).

**Comment by jimbo** [comment]
Thanks for the quick responses. If we conclude that this isn't suitable as an option, I'll certainly investigate the plugin route.

> I think it's a clear candidate for a plugin, because this option doesn't have a clear and unambiguous finished state, and, I believe, won't be used by majority of stylelint's users.

I saw that in the rules, but I don't think I understand what it means for a rule/option to have a "finished state". Could you clarify what the state of a rule/option is, and how that state might be construed as clear or unclear?

> It took me a bit to get my head around it, and I'm not sure I understand it completely yet. Am I right in thinking that the way the `selector-max-specificity` currently works is in line with with the spec, and what you're proposing is some adaptation of that?

Apologies for not making it clearer; I'll try to explain with code.

Looking at the [existing rule](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-max-specificity/index.js#L52-L65), the rule reports a violation based on a single condition:

```js
if (specificity.compare(resolvedSelector, maxSpecificityArray) === 1) {
  report()
}
```

Internally, the `specificity` module's [`compare()`](https://github.com/keeganstreet/specificity/blob/master/specificity.js#L166-L220) function uses a local function, [`calculateSingle()`](https://github.com/keeganstreet/specificity/blob/master/specificity.js#L27-L164), to calculate the specificity of the selector, best expressed as an array of the format `[0, 1, 3, 1]`. It then calculates the *cumulative* specificity of that array and the maximum array, according to the CSSWG spec, and compares it to the *cumulative* specificity of the maximum array, returning a value between `-1` and `1`.

I'm proposing to—optionally—avoid the `compare()` function and its calculation of cumulative specificity. Instead, I'm proposing to use `calculate()` and compare specificity arrays directly.

```js
// If the `granular` option is enabled, check each parameter individually
if (isGranular) {
  const { specificityArray } = specificity.calculate(resolvedSelector).shift()

  if (specificityArray.some((parameter, index) => parameter > maxSpecificityArray[index])) {
    report()
  }
// Otherwise check total specificity
} else if (specificity.compare(resolvedSelector, maxSpecificityArray) === 1) {
  report()
}
```

The issue here isn't how cumulative specificity is being calculated; it's that it isn't an ideal metric for authoring CSS. I don't have usage statistics, but I think a typical CSS codebase would consist of many zero-id selectors and a few one-id selectors. Some codebases avoid id selectors entirely; many cannot.

Accordingly, a common configuration for `selector-max-specificity` would be something like `"1,x,y"`. In other words, your selectors may include up to one id selector. However, this configuration makes the rule powerless to govern zero-id selectors, which I think would comprise the majority of a typical codebase. With `"1,0,0"`, **literally any selector that doesn't include an id will pass the rule**, regardless of how extreme its specificity is. (Correct me if I'm wrong.)

When I configure a max specificity of `"1,3,1"`, what I'm really saying is that I want my selectors to contain at most one id, at most three classes, *and* at most one type. If I write a zero-id selector such as `.lorem.ipsum .dolor.sit .amet`, I want the rule to consider each piece separately. Otherwise, I might as well configure the rule with `"1,0,0"`, because it makes no difference.

> If so, then I agree with @hudochenkov that this is a perfect candidate for a plugin. Plugins are where language extensions live. You're granular name makes sense to me, and so you may want to call the plugin stylelint-selector-max-granular-specificity, or something similar.

Thanks for referring me to that list. I'll look into other plugins and see what they're doing. If I made a plugin, I suppose I could duplicate the existing rule, change the one line, and use it instead of the existing rule.

**Comment by jeddy3** [comment]
@jimbo Many thanks for taking the time to explain this in more depth. It's much clearer now.

> With "1,0,0", literally any selector that doesn't include an id will pass the rule, regardless of how extreme its specificity is. (Correct me if I'm wrong.)

You are not wrong!

> When I configure a max specificity of "1,3,1", what I'm really saying is that I want my selectors to contain at most one id, at most three classes, and at most one type. If I write a zero-id selector such as .lorem.ipsum .dolor.sit .amet, I want the rule to consider each piece separately.

That's how I always thought this rule worked, but now I see that that is not the case at all. Whoops!

So, from [the spec](https://www.w3.org/TR/selectors4/#specificity):

> Specificities are compared by comparing the three components in order: the specificity with a larger A value is more specific; if the two A values are tied, then the specificity with a larger B value is more specific; if the two B values are also tied, then the specificity with a larger c value is more specific; if all the values are tied, the two specifities are equal.

I guess the confusion comes down to the naming of the rule... specificity is normally used to determine which selector _is more specific_ than another. In `selector-max-specificity` we're using the primary option (i.e. the `max` value) as the specificity of an imaginary selector. The other `-max-` are more obvious than this.

I agree with you that a granular comparison is the more useful. However, I'm wondering whether this is the rule where we should do it. For example, from the spec:

> A selector's specificity is calculated as follows:
> - count the number of ID selectors in the selector (= A)
> - count the number of class selectors, attributes selectors, and pseudo-classes in the selector (= B)
> - count the number of type selectors and pseudo-elements in the selector (= c)

Elsewhere in stylelint we have `-max-` rules that target _very specific things_ e.g. `declaration-block-single-line-max-declarations`, `selector-max-compound-selectors` and `number-max-precision`. If we had granular support for specifying the maximums of "A", "B" and "c" then we're going to be munging together class selectors, attributes selectors, and pseudo-classes for "B" and type selectors and pseudo-elements for "c". I think munging these things together makes sense _when comparing the specificity of selectors_, but not when _specifying a maximum of something within a codebase_.

As such, I'm wondering whether we keep the `selector-max-specificity` rule as it is and instead add a series of `selector-max-*` rules e.g.:

- `selector-max-type`
- `selector-max-pseudo-element`

...

- `selector-max-class`
- `selector-max-attribute`
- `selector-max-pseudo-class`

...

- `selector-max-id`

For example, I suspect people might want to independently apply limitations to class, attribute and  pseudo-class selectors:

```json
// Allow one class selector, no attribute selectors and any number of pseudo-class selectors */
{  
  "selector-max-class": 1,
  "selector-max-attribute": 0,
}
```

This would not be possible if we only add granular checking to the `selector-max-specificity` rule.

However, such rules would make existing `selector-no-*` rules redundant, but I think we can cross the bridge later.

@jimbo What do you think? Would such rules cover of your use case? They would certainly be more in line with the stylelint way of doing things.

**Comment by jimbo** [comment]
Thanks for the detailed response, @jeddy3. I think we're on the same page now.

> If we add granular support for specifying the maximums of "A", "B" and "c" then we're going to be munging together class selectors, attributes selectors, and pseudo-classes for "B" and type selectors and pseudo-elements for "c". I think munging these things together makes sense when comparing the specificity of selectors, but not when specifying a maximum of something within a codebase.

> For example, I suspect people might want to independently apply limitations to class, attribute and pseudo-class selectors

You're absolutely right. The only reason to put tags and pseudo-elements in the same bin is when calculating specificity. If we're going to be limiting their count instead, it makes sense to track and limit them separately. Likewise for the various other types of selectors.

Without going into more inline quotes, I agree that the better approach here is probably to add a set of new rules (and leave the existing rule alone). I do think the new rules you've outlined would cover my use case and many more, and I'd be happy to take a shot at writing them. I'll spike on it for a bit.

👍

**Comment by jeddy3** [comment]
> I think we're on the same page now.

Good stuff!

> I do think the new rules you've outlined would cover my use case and many more, and I'd be happy to take a shot at writing them. 

Excellent. I suggest we start with `selector-max-class` as that won't overlap with any existing rules i.e. there are `selector-no-id`, `selector-no-attribute` and `selector-no-type` rules, but no `selector-no-class` rule. Once we have `selector-max-class` nail down we can discuss the merits of the other rules in separate issues.

I suggest we [_scope_](https://stylelint.io/user-guide/about-rules/#about-rule-names) the rule to each _selector_ i.e. the tally is reset for each _selector_ within a _selector list_. For example:

```json
{ 
  "selector-max-class": 1
}
```

Will _not_ flag any warnings against:

```css
.class1,
.class2 {}
```

- Name: `selector-max-class`
- Primary options: `int`
- Secondary options: none
- Message: `Expected no more than ${max} class selectors`
- Section: "Selector"

@stylelint/core I'll label this up. Shout if you see any issues with this new rule. Otherwise, I think we're good to let @jimbo crack on.

**Comment by jeddy3** [comment]
Done in #2474
